core/domain/mailchimp_services.py

```diff
@@ -0,0 +1,172 @@
+
+from core.platform import models
+from mailchimp3 import MailChimp
+from mailchimp3.mailchimpclient import MailChimpError
+
+import feconf
+import hashlib
+import python_utils
+
+(user_models,) = models.Registry.import_models([models.NAMES.user])
+
+def _get_mailchimp_class():
+    """Returns the mailchimp api class. This is separated into a separate 
+    function to facilitate testing. 
+
+    NOTE: No other functionalities should be added to this function.
+    """
+    if not feconf.MAILCHIMP_API_KEY:
+        raise Exception('Mailchimp API key is not available.')
+
+    if not feconf.MAILCHIMP_USERNAME:
+        raise Exception('Mailchimp username is not set.')
+
+    return MailChimp(
+        mc_api=feconf.MAILCHIMP_API_KEY, mc_user=feconf.MAILCHIMP_USERNAME)
+
+
+def get_subscriber_hash(email):
+    """Returns Mailchimp subscriber hash from email.
+    
+    Args:
+        email: str. The email ID of the user.
+    
+    Raises:
+        Exception. Invalid type for email. Expecetd string.
+    
+    Returns:
+        str. The subscriber hash corresponding to input email.
+    """
+    if not isinstance(email, python_utils.BASESTRING):
+        raise Exception(
+            'Invalid type for email. Expected string, received %s' % email) 
+    md5_hash = hashlib.md5()
+    md5_hash.update(email)
+    return md5_hash.hexdigest()
+
+
+def permanently_delete_user_from_list(user_email):
+    """Permanently deletes user from mailchimp list. 
+    
+    NOTE: This should only be called from the wipeout service since once a user 
+    is permanently deleted from mailchimp, they cannot be programmatically added 
+    back via their API (the user would have to manually resubscribe back).
+
+    Args:
+        user_email: str. Email id of the user.
+    
+    Raises:
+        Exception: Mailchimp API key is not available.
+        Exception: Mailchimp username is not set.
+    """
+    client = _get_mailchimp_class()
+    
+    try:
+        member_details = client.lists.members.get(
+            list_id=feconf.MAILCHIMP_AUDIENCE_ID, 
+            subscriber_hash=get_subscriber_hash(user_email))
+        client.lists.members.delete_permanent(
+            list_id=feconf.MAILCHIMP_AUDIENCE_ID, 
+            subscriber_hash=get_subscriber_hash(user_email))
+    except MailChimpError as error:
+        # Ignore if the error corresponds to "User does not exist"
+        if error[0]['status'] != 404:
+            raise Exception(error[0]['detail'])
+
+
+def add_or_update_mailchimp_user_status(
+        user_id, user_email, can_receive_email_updates):
+    """Adds a new user to the mailchimp mailing list as well as enables their
+    preference in the UserEmailPreferencesModel.
+
+    Note: The UserEmailPreferencesModel for the user should already be created
+    before this function is called.
+
+    Args:
+        user_id: str. The ID of the user.
+        user_email: str. The email ID of the user.
```

Comment:
Let's not use the term "email ID", just "email" will do. This is to avoid confusion with user_id.

Reply:
Ditto above about using email ID.